File: day-07/handy_haversacks.py
import collections, re

input_file = 'day-07/input.txt'

def calculate_answer(input, answer_func):
  parents = collections.defaultdict(set)
  children = collections.defaultdict(list)
  for line in input.splitlines():
    parent_colour = re.match(r'(\w+ \w+) bags contain', line)[1]
    for quantity, colour in re.findall(r'(\d+) (\w+ \w+) bags?[,.]', line):
      parents[colour].add(parent_colour)
      children[parent_colour].append((int(quantity), colour))
  return answer_func(parents, children)

def find_ancestors(node, parents):
  ancestors = set()
  for parent in parents[node]:
    ancestors.add(parent)
    ancestors = ancestors.union(find_ancestors(parent, parents))
  return ancestors

def part1(input):
  gold_ancestors = lambda parents, children: len(find_ancestors('shiny gold', parents))
  return calculate_answer(input, gold_ancestors)

def count_children(node, children):
  descendants = 0
  for quantity, child in children[node]:
    descendants += quantity
    descendants += (quantity * count_children(child, children))
  return descendants

def part2(input):
  gold_contents = lambda parents, children: count_children('shiny gold', children)
  return calculate_answer(input, gold_contents)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-07/handy_haversacks.go
package main

import (
	"fmt"
	"io/ioutil"
	"regexp"
	"strconv"
	"strings"
)

const input_file string = "./day-07/input.txt"

type Bag struct {
	name     string
	contents map[*Bag]int
}

func (b Bag) Find(target string) bool {
	for c := range b.contents {
		if c.name == target || c.Find(target) {
			return true
		}
	}
	return false
}

func (b Bag) Count() int {
	count := 0
	for c, quantity := range b.contents {
		count += quantity + (quantity * c.Count())
	}
	return count
}

func ParseInput(input string) map[string]*Bag {
	bags := make(map[string]*Bag)

	parentMatcher := regexp.MustCompile(`^(\w+ \w+) bags contain (.*)$`)
	childMatcher := regexp.MustCompile(`^(\d+) (\w+ \w+) bag[s]?[.]?$`)

	lines := strings.Split(strings.TrimSpace(input), "\n")
	for _, line := range lines {
		matches := parentMatcher.FindStringSubmatch(line)
		bagName := matches[1]
		bag, ok := bags[bagName]
		if !ok {
			bag = &Bag{name: bagName, contents: make(map[*Bag]int)}
			bags[bagName] = bag
		}

		for _, contents_string := range strings.Split(strings.TrimRight(matches[2], "."), ", ") {
			// fmt.Printf("Contents %s\n", contents_string)
			if contents_string == "no other bags" {
				continue
			}
			contents := childMatcher.FindStringSubmatch(contents_string)
			quantity, err := strconv.Atoi(contents[1])
			if err != nil {
				panic(err)
			}
			childBagName := contents[2]
			childBag, ok := bags[childBagName]
			if !ok {
				childBag = &Bag{name: childBagName, contents: make(map[*Bag]int)}
				bags[childBagName] = childBag
			}
			bag.contents[childBag] = quantity
		}
	}

	return bags
}

func Part1(input string) int {
	count := 0
	bags := ParseInput(input)
	for _, b := range bags {
		if b.Find("shiny gold") {
			count++
		}
	}
	return count
}

func Part2(input string) int {
	bags := ParseInput(input)
	return bags["shiny gold"].Count()
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
